File: core/src/main/java/no/elg/infiniteBootleg/world/time/WorldTime.java
package no.elg.infiniteBootleg.world.time;

import com.badlogic.gdx.graphics.Color;
import no.elg.infiniteBootleg.util.Util;
import no.elg.infiniteBootleg.world.World;
import org.jetbrains.annotations.NotNull;

public class WorldTime {

  /**
   * How many degrees the time light should have before triggering sunset/sunrise. This will happen
   * from {@code -TWILIGHT_DEGREES} to {@code +TWILIGHT_DEGREES}
   */
  public static final float TWILIGHT_DEGREES = 18;

  /** The moment the sun is parallel to the horizon and is rising */
  public static final float SUNRISE_TIME = 180 + TWILIGHT_DEGREES;
  /** When the world begins to lighten up */
  public static final float DAWN_TIME = 180;

  /** The moment the sun is parallel to the horizon and is sinking */
  public static final float SUNSET_TIME = 0;
  /** When there is no more ambient light from the sun */
  public static final float DUSK_TIME = SUNSET_TIME + TWILIGHT_DEGREES;

  /** Middle of the day */
  public static final float MIDDAY_TIME = 270;

  /** Middle of the night */
  public static final float MIDNIGHT_TIME = 90;

  @NotNull private final World world;
  @NotNull private final Color baseColor = new Color(Color.WHITE);

  private volatile float time = SUNRISE_TIME;
  private volatile float timeScale = 1f;

  public WorldTime(@NotNull World world) {
    this.world = world;
  }

  /**
   * Calculate how bright the sky should be. During the night the value will always be {@code 0},
   * during twilight (ie from {@code 360} to {@code 360-}{@link WorldTime#TWILIGHT_DEGREES} and
   * {@code 180+}{@link WorldTime#TWILIGHT_DEGREES} to {@code 180-}{@link
   * WorldTime#TWILIGHT_DEGREES}) the light will change. During daytime the value will always be 1
   *
   * <p>The time used will be the current world time ie {@link #getTime()}
   *
   * @return A brightness value between 0 and 1 (both inclusive)
   */
  public float getSkyBrightness() {
    return getSkyBrightness(time);
  }

  /**
   * Calculate how bright the sky should be. During the night the value will always be {@code 0},
   * during twilight (ie from {@code 360} to {@code 360-}{@link WorldTime#TWILIGHT_DEGREES} and
   * {@code 180+}{@link WorldTime#TWILIGHT_DEGREES} to {@code 180-}{@link
   * WorldTime#TWILIGHT_DEGREES}) the light will change. During daytime the value will always be 1
   *
   * @param time The time to calculate
   * @return A brightness value between 0 and 1 (both inclusive)
   */
  public float getSkyBrightness(float time) {
    float dir = Util.normalizedDir(time);

    if (dir >= DAWN_TIME && dir <= SUNRISE_TIME) {
      return (dir - DAWN_TIME) / TWILIGHT_DEGREES;
    } else if (dir >= SUNSET_TIME && dir <= DUSK_TIME) {
      return 1 - dir / TWILIGHT_DEGREES;
    } else if (dir <= 180) {
      return 0;
    } else if (dir > 180) {
      return 1;
    }
    return 1;
  }

  public float normalizedTime() {
    return Util.normalizedDir(time);
  }

  public float getTimeScale() {
    return timeScale;
  }

  public void setTimeScale(float timeScale) {
    this.timeScale = timeScale;
  }

  public float getTime() {
    return time;
  }

  public void setTime(float time) {
    this.time = time;
  }

  public Color getBaseColor() {
    return baseColor;
  }

  public World getWorld() {
    return world;
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/core/world/WorldTime.kt
package no.elg.infiniteBootleg.core.world

import com.badlogic.gdx.graphics.Color
import no.elg.infiniteBootleg.core.util.Util
import no.elg.infiniteBootleg.core.world.world.World

class WorldTime(val world: World) {
  val baseColor = Color(Color.WHITE)

  @Volatile
  var time = SUNRISE_TIME

  @Volatile
  var timeScale = DEFAULT_TIME_SCALE

  /**
   * Calculate how bright the sky should be. During the night the value will always be `0`,
   * during twilight (ie from `360` to `360-`[WorldTime.TWILIGHT_DEGREES] and
   * `180+`[WorldTime.TWILIGHT_DEGREES] to `180-`[WorldTime.TWILIGHT_DEGREES]) the light will change. During daytime the value will always be 1
   *
   *
   * The time used will be the current world time ie [getTime]
   *
   * @return A brightness value between 0 and 1 (both inclusive)
   */
  val skyBrightness: Double
    get() = getSkyBrightness(time)

  /**
   * Calculate how bright the sky should be. During the night the value will always be `0`,
   * during twilight (ie from `360` to `360-`[WorldTime.TWILIGHT_DEGREES] and
   * `180+`[WorldTime.TWILIGHT_DEGREES] to `180-`[WorldTime.TWILIGHT_DEGREES]) the light will change. During daytime the value will always be 1
   *
   * @param time The time to calculate
   * @return A brightness value between 0 and 1 (both inclusive)
   */
  fun getSkyBrightness(time: Double = this.time): Double =
    atTime(
      time = time,
      day = { 1.0 },
      dusk = { 1.0 - (it - SUNSET_TIME) / (DUSK_TIME - SUNSET_TIME) },
      night = { 0.0 },
      dawn = { (it - DAWN_TIME) / (SUNRISE_TIME - DAWN_TIME) }
    )

  fun timeOfDay(time: Double = this.time): String =
    atTime(
      time = time,
      day = { "Day" },
      dusk = { "Dusk" },
      night = { "Night" },
      dawn = { "Dawn" }
    )

  inline fun <T> atTime(
    time: Double = this.time,
    day: (Double) -> T,
    dusk: (Double) -> T,
    night: (Double) -> T,
    dawn: (Double) -> T
  ): T {
    val dir = Util.normalizedDir(time)
    return when {
      // The sun has not set yet
      dir < SUNSET_TIME -> day(dir)

      // As the sun sinks the ambient light fades
      dir in SUNSET_TIME..DUSK_TIME -> dusk(dir)

      // Night between dusk and dawn
      dir in DUSK_TIME..DAWN_TIME -> night(dir)

      // As the sun returns the ambient light increases
      dir in DAWN_TIME..SUNRISE_TIME -> dawn(dir)

      // Day between sunrise and sunset, no need for an explicit check here
      else -> day(dir)
    }
  }

  fun normalizedTime(): Double = Util.normalizedDir(time)

  companion object {
    const val DEFAULT_TIME_SCALE = 0.033 // roughly 1/30

    /**
     * How many degrees the time light should have before triggering sunset/sunrise. This will happen
     * from `-TWILIGHT_DEGREES` to `+TWILIGHT_DEGREES`
     */
    const val TWILIGHT_DEGREES = 18.0

    /**
     * The moment the sun is parallel to the horizon and is rising
     */
    const val SUNRISE_TIME = 180.0 - TWILIGHT_DEGREES

    /**
     * When the world begins to lighten up
     */
    const val DAWN_TIME = SUNRISE_TIME - TWILIGHT_DEGREES

    /**
     * The moment the sun is parallel to the horizon and is sinking
     */
    const val SUNSET_TIME = 0.0 + TWILIGHT_DEGREES

    /**
     * When there is no more ambient light from the sun
     */
    const val DUSK_TIME = SUNSET_TIME + TWILIGHT_DEGREES

    /**
     * Middle of the day
     */
    const val MIDDAY_TIME = 270.0

    /**
     * Middle of the night
     */
    const val MIDNIGHT_TIME = 90.0
  }
}
